localstack/mock/infra.py

```diff
@@ -358,27 +361,33 @@ def update_apigateway(method, path, data, headers, response=None, return_forward
             # forward records to our main kinesis stream
             # TODO check whether the target of this API method is 'kinesis'
             headers = aws_stack.mock_aws_request_headers(service='kinesis')
-            headers['X-Amz-Target'] = 'Kinesis_20131202.PutRecords'
+            headers['X-Amz-Target'] = KINESIS_ACTION_PUT_RECORDS
             result = common.make_http_request(url=TEST_KINESIS_URL,
                 method='POST', data=new_request, headers=headers)
             return 200
         return True
 
 
-def update_kinesis(method, path, data, headers, response=None, return_forward_info=False):
+def update_kinesis(method, path, data, headers, response=None, return_forward_info=False, return_errors=0):
```

Comment:
ok, so this is how this method works (it's a bit unintuitive unfortunately).

We have a proxy that sits in front of Kinesalite and intercepts all Kinesis API calls. When we create the proxy we pass a pointer to the `update_kinesis` function: https://github.com/atlassian/localstack/blob/f4e5151d80d62196e4b95fe0050a4a2f483bae18/localstack/mock/infra.py#L108.

For each API call, the proxy will call this method (`update_kinesis`) twice:
1) `update_kinesis(..., return_forward_info=True)` to return whether or not to forward the request to Kinesalite or drop it. This happens here: https://github.com/atlassian/localstack/blob/f4e5151d80d62196e4b95fe0050a4a2f483bae18/localstack/mock/generic_proxy.py#L79
2) `update_kinesis(..., return_forward_info= False)` to inform this method about the result. This second call only happens if the previous call has returned `True`. The second call happens from here: https://github.com/atlassian/localstack/blob/f4e5151d80d62196e4b95fe0050a4a2f483bae18/localstack/mock/generic_proxy.py#L93

So, the first call happens *before* the actual Kinesalite invocation to allow this method to decide whether the request should be forwarded or not. The second call happens *after* the actual Kinesalite invocation to allow the method to process the response returned by Kinesalite.

Now, unfortunately things become a bit more complicated. :) The first call of this listener function allows not only to return a boolean result (`True`/`False`) whether to forward the message, but additionally also allows to forward an integer value (e.g., 200) which indicates "don't forward the message to the target, but return a 200 response to the client right away". (for instance, see here: https://github.com/atlassian/localstack/blob/f4e5151d80d62196e4b95fe0050a4a2f483bae18/localstack/mock/infra.py#L348)

Needless to say, this method is way too overloaded and the functionality is not properly documented. I'd be happy to contribute and refactor/document the code a bit to make it more easily understandable.

So, I think what we need to do in this PR is to react to the *first* call of this method (if `return_forward_info=True`). In case we want to inject an error, we could return a corresponding 500 `Response` object and have the proxy return it to the client. We should also make sure that the original request should not be forwarded to Kinesalite, as this could create weird behavior (500 error, yet the record becomes available on the stream).